services/core/users/src/main/java/org/opfab/users/controllers/UsersController.java

```diff
@@ -193,6 +204,62 @@ public Void deleteUser(HttpServletRequest request, HttpServletResponse response,
         return null;
     }
 
+
+
+    @Override
+    public User synchronizeWithToken(HttpServletRequest request, HttpServletResponse response) throws Exception {
+        User user = this.extractUserFromJwtToken(request);
+
+        checkAndsetUserLogin(user);
+
+        String login = user.getLogin();
+
+        if (groupsProperties.getMode() == GroupsMode.JWT) {
+            List<String> missingGroups = user.getGroups().stream()
+                    .filter(groupId -> this.groupRepository.findById(groupId).isEmpty()).collect(Collectors.toList());
+            if (!missingGroups.isEmpty()) {
+                missingGroups.forEach(id -> log.warn("Group id from token not found in db: {}", id));
+                List<String> goodGroups = user.getGroups();
+                goodGroups.removeAll(missingGroups);
+                user.setGroups(goodGroups);
+            }
+        }
+
+        if (jwtProperties.gettingEntitiesFromToken) {
+            List<String> missingEntities = user.getEntities().stream()
+                    .filter(entityId -> this.entityRepository.findById(entityId).isEmpty())
+                    .collect(Collectors.toList());
+            if (!missingEntities.isEmpty()) {
+                missingEntities.forEach(id -> log.warn("Entity id from token not found in db: {}", id));
+                List<String> goodEntities = user.getEntities();
+                goodEntities.removeAll(missingEntities);
+                user.setEntities(goodEntities);
+            }
+        }
+
+        UserData existingUser = userRepository.findById(login).orElse(null);
+
+        if (existingUser == null) {
+            log.debug(String.format(USER_CREATED, login));
+            userService.createUser(user);
+        } else {
+            boolean updatedFromToken = false;
+            if (groupsProperties.getMode() == GroupsMode.JWT) {
+                updatedFromToken = !existingUser.getGroupSet().equals(new HashSet<String>(user.getGroups()));
+            }
+            if (!updatedFromToken && jwtProperties.gettingEntitiesFromToken) {
+                updatedFromToken = !new HashSet<String>(existingUser.getEntities()).equals(new HashSet<String>(user.getGroups()));
```

Comment:
it shall be :  
updatedFromToken = !new HashSet<String>(existingUser.getEntities()).equals(new HashSet<String>(user.**getEntities())**); ?